### PR TITLE
Feature coalescing

### PR DESCRIPTION
**Pull request by webmiche**
Updated version of the PR after rebasing

### CODE REVIEW

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace mlir {
+
+// coalesces a set according to the "integer set coalescing" by sven
+// verdoolaege.
+//
+// Coalescing task two convex BasicSets and tries to figure out, whether the
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Coalesce.h]: nit: "Coalescing _takes_..."?

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// convex hull of those two BasicSets is the same integer set as the union of
+// those two BasicSets and if so, tries to come up with a BasicSet corresponding
+// to this convex hull.
+PresburgerSet coalesce(PresburgerSet &set);
+
+void dump(const ArrayRef<int64_t> cons);
+
+// compare two constraints and gives true, even if they are stretched
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Coalesce.h]: Please elaborate here on what "stretched" means.

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+bool sameConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2);
+
+// tries to find a constraint, that is a linear combination of valid and
+// invalid, and touches bs/is REDUNDANT in the most limited way possible
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Coalesce.h]: Please elaborate here on what "most limited way possible" means.

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                                           SmallVectorImpl<int64_t> &invalid);
+
+// combine two constraints c1 >= 0 and c2 >= 0 with the ratio n/d as -n c1 + d
+// c2 >= 0
+SmallVector<int64_t, 8>
+combineConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2, Fraction &ratio);
+
+// takes a BasicSet bs, a constraint ineq of that basicSet and the vector cut of
+// constraints, that were typed as cutting bs. Computes wether the part of ineq,
+// that lies within bs, is redundant for all constraints of cut
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Coalesce.h]: nit: Should this be, "compute whether _all constraints of cut are redundant_ for the part of ineq that lies within bs"?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+// respectively. adj_ineq is for any constraint, that is adjacent to the other
+// polytope. t is for any constraint, that is part of an equality constraint and
+// adjacent to the other polytope.
+// TODO: find better name than t
+// TODO: possibly change up structure of Info, if simplex manages to classify
+// adjacent to equality
+struct Info {
+  SmallVector<ArrayRef<int64_t>, 8> redundant;
+  SmallVector<ArrayRef<int64_t>, 8> cut;
+  Optional<ArrayRef<int64_t>> adj_ineq;
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: nit: LLVM uses camelCase for variables, not snake_case.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+// computes the complement of t
+// i.e. for a given constraint t(x) >= 0 returns -t(x) -1 >= 0
+SmallVector<int64_t, 8> complement(ArrayRef<int64_t> t);
+
+// shifts t by amount
+// i.e. for a given constraint t(x) >= 0 return t(x) + amount >= 0
+void shift(SmallVectorImpl<int64_t> &t, int amount);
+
+// add eq as two inequalities to target
+void addAsIneq(const ArrayRef<ArrayRef<int64_t>> eq,
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: Maybe indicate what is being added in the function name, e.g., `addEqualitiesAsInequalities`.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+void addInequalities(FlatAffineConstraints &bs,
+                     const SmallVector<ArrayRef<int64_t>, 8> &inequalities);
+
+// only gets called by classify
+// classifies all constraints into redundant, cut or adj_ineq according to the
+// ineqType that the simplex returns
+//
+// returns true if it has neither encountered a separate constraint nor more
+// than one adjacent inequality
+bool classify_ineq(Simplex &simp,
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: LLVM style would have this be `classifyIneq`

File: mlir/include/mlir/Analysis/Presburger/Simplex.h [hunk 8]
```diff
@@ -128,6 +128,8 @@ class Simplex {
 public:
   enum class Direction { Up, Down };
 
+  enum class IneqType { REDUNDANT, SEPARATE, CUT, ADJ_EQ, ADJ_INEQ };
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Simplex.h]: LLVM style would use UpperCamelCase for enum values

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+                   const SmallVector<ArrayRef<int64_t>, 8> &constraints,
+                   Info &info);
+
+// classifies all constraints into redundant, cut, adj_ineq or t, where t stands
+// for a constraint adjacent to a the other polytope
+//
+// returns true if it has not encountered a separate constraint or more than one
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: A little explanation of why this is the return value would be nice.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 7 above)
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: A little explanation of why this is the return value would be nice. Also below for `classify`

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h (hunk 4 above)
Comment (superty) [reply mlir/include/mlir/Analysis/Presburger/Coalesce.h]: Please write that by "the part of ineq that lies within bs", it means the part of bs that satisfies ineq with equality, or something like that, if that is what it is supposed to do.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 10]
```diff
@@ -0,0 +1,909 @@
...
+  for (size_t i = 0; i < cut.size(); i++) {
+    if (simp.ineqType(cut[i]) != Simplex::IneqType::REDUNDANT) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void addAsIneq(ArrayRef<ArrayRef<int64_t>> eq,
+               SmallVectorImpl<ArrayRef<int64_t>> &target) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: Would it be possible to use a `MutableArrayRef` of `SmallVector`s, and avoid the pointer stuff below?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+  for (size_t i = 0; i < eq.size(); i++) {
+    ArrayRef<int64_t> curr = eq[i];
+    target.push_back(curr);
+    // TODO: fix this memory leak
+    SmallVector<int64_t, 8> *inverted = new SmallVector<int64_t, 8>();
+    for (size_t j = 0; j < curr.size(); j++) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: range-based for loop?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+    if (simp.ineqType(cut[i]) != Simplex::IneqType::REDUNDANT) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void addAsIneq(ArrayRef<ArrayRef<int64_t>> eq,
+               SmallVectorImpl<ArrayRef<int64_t>> &target) {
+  for (size_t i = 0; i < eq.size(); i++) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: range-based for loop?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 13]
```diff
@@ -0,0 +1,909 @@
...
+  addCoalescedBasicSet(basicSetVector, i, j, new_set);
+  return true;
+}
+
+bool mlir::containedFacet(ArrayRef<int64_t> ineq,
+                          const FlatAffineConstraints &bs,
+                          const SmallVector<ArrayRef<int64_t>, 8> &cut) {
+  Simplex simp(bs);
+  simp.addEquality(ineq);
+  for (size_t i = 0; i < cut.size(); i++) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: range-based for loop?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 14]
```diff
@@ -0,0 +1,909 @@
...
+// returns all Inequalities of a BasicSet as a SmallVector of ArrayRefs
+void getBasicSetInequalities(const FlatAffineConstraints &bs,
+                             SmallVector<ArrayRef<int64_t>, 8> &ineqs) {
+  for (unsigned k = 0; k < bs.getNumInequalities(); k++) {
+    ineqs.push_back(bs.getInequality(k));
+  }
+}
+
+PresburgerSet mlir::coalesce(PresburgerSet &set) {
+  PresburgerSet new_set(set.getNumDims(), set.getNumSyms());
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: LLVM uses camelCase instead of snake_case

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 15]
```diff
@@ -0,0 +1,909 @@
...
+                 const FlatAffineConstraints &bs);
+
+// adds a FlatAffineConstraints and removes the sets at i and j.
+void addCoalescedBasicSet(
+    SmallVectorImpl<FlatAffineConstraints> &basicSetVector, unsigned i,
+    unsigned j, const FlatAffineConstraints &bs);
+
+// compute the cut case and return whether it has worked.
+//
+// The cut case is the case, for which a polytope only has REDUNDANT and CUT
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: As far as I understand, the sloped cut constraint in the left diagram is not totally contained in the other polytope? What am I missing? Please elaborate in the documentation.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 16]
```diff
@@ -0,0 +1,909 @@
...
+// compute the non-pure adj_ineq case and return whether it has worked.
+// Constraint t is the adj_ineq.
+//
+// In the non-pure adj_ineq case, one of the polytopes is like an extension of
+// the other one. This can be computed, by inverting the adj_ineq and checking
+// whether all constraints are stll valid for this new polytope.
+//   ____          ____
+//  |    \        |    \
+//  |     \   ==> |     \
+//  |_____|_      |______\
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: The second polytope is not clearly visible in this diagram. Please write a line pointing out where it is.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 17]
```diff
@@ -0,0 +1,909 @@
...
+  // the optimum is < -2, the polytopee doesn't stick too much out of the other
+  // one.
+  for (size_t k = 0; k < cut.size(); k++) {
+    ArrayRef<int64_t> curr = cut[k];
+    auto result = simp.computeOptimum(Simplex::Direction::Down, curr);
+    if (!result) {
+      return false;
+    }
+    auto res = result.getValue();
+    if (res.num <= -2 * res.den) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: Might be cleaner to write it as `res <= Fraction(-2, 1)`.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 18]
```diff
@@ -0,0 +1,909 @@
...
+  SmallVector<int64_t, 8> complement;
+  for (size_t k = 0; k < t.size() - 1; k++) {
+    complement.push_back(-t[k]);
+  }
+  complement.push_back(-t.back() - 1);
+  return complement;
+}
+
+// helperfuncton to convert arrayRefs to SmallVectors
+static SmallVector<int64_t, 8> arrayRefToSmallVector(ArrayRef<int64_t> ref) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: You can use `llvm::to_vector`.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 19]
```diff
@@ -0,0 +1,909 @@
...
+      return false;
+    }
+  }
+  return true;
+}
+
+bool mlir::sameConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2) {
+  Fraction ratio(0, 1);
+  if (c1.size() != c1.size()) {
+    return false;
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: Should this rather be an assertion failure? Can this ever happen normally?

File: mlir/lib/Analysis/Presburger/Coalesce.cpp [hunk 20]
```diff
@@ -0,0 +1,909 @@
...
+  }
+  // if c1 = a*c2, this iterates over the vector trying to find a as soon as
+  // possible and then comparing with a.
+  for (size_t i = 0; i < c1.size(); i++) {
+    // TODO: is ratio.num == 0 really sufficient for checking whether fraction
+    // was set yet
+    if (c2[i] != 0 && ratio.num == 0) {
+      ratio.num = c1[i];
+      ratio.den = c2[i];
+      if (ratio.den * c1[i] != ratio.num * c2[i]) {
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Coalesce.cpp]: This check is redundant, right?

**Review by superty — COMMENTED** [review]
Thanks for this PR. Some general comments:

- Especially since we plan to upstream this eventually, we should follow LLVM style, and have sufficiently detailed documentation, so I have left several comments to that effect.

- Is it possible to change the const SmallVector references to ArrayRefs? The SmallVectors of ArrayRefs included.

- Try to use range-based for loops when possible.

- Use camelCase instead of snake_case for variable and function names.

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h (hunk 1 above)
Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Coalesce.h]: yes, thx

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h (hunk 2 above)
Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Coalesce.h]: I rewrote this to:
compare two constraints and give true if they are equal. Can also handle cases in which for some integers a and b, c1 = a/b * c2. This is the same constraint but stretched, which doesn't influence it's hyperplane.

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h (hunk 3 above)
Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Coalesce.h]: I rewrote this to:
rotates invalid around valid, until it becomes redundant. It does this by adding the smallest multiple of valid to invalid, such that the result is redundant.

File: mlir/include/mlir/Analysis/Presburger/Coalesce.h (hunk 4 above)
Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Coalesce.h]: Yes, that's what it is supposed to do. I rewrote this.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 5 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: rewrote everything, _ only occurs in types and vector instructons now.
And in the enumeration of IneqType. We might have to change it there too (ADJ_INEQ --> ADJINEQ)

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 16 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I added:
In the example below, Polytope A is the big one on the left, Polytope B is the small line at the right side. that gets included after coalescing.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 6 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 7 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 5 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I just saw your comment below, nevermind the last line ;)

File: mlir/include/mlir/Analysis/Presburger/Simplex.h (hunk 8 above)
Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Simplex.h]: made ADJ_INEQ --> ADJINEQ and ADJ_EQ --> ADJ_EQ

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 9 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I added:
returns true if it has neither encountered a separate constraint nor more than one adjacent inequality. This has a similar idea as short-circuiting behind it. The moment either a separate or two adjacent constraints are encountered, the polytope cannot be coalesced anymore, so we can move to the next tuple.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 7 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I added (same as below):
returns true if it has neither encountered a separate constraint nor more than one adjacent inequality. This has a similar idea as short-circuiting behind it. The moment either a separate or two adjacent constraints are encountered, the polytope cannot be coalesced anymore, so we can move to the next tuple.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 10 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: The pointer stuff is needed, to make all constraints become `ArrayRef`s, which suffices for the rest of the program and AFAIU is much faster than using something that is mutable like `MutableArrayRef`or something similar.

I didn't actually give much thought into swapping `SmallVectorImpl` to `MutableArrayRef` as I felt like we will need to redo a lot of those things as soon as the Constraint class and the BasicSet class have landed.

I added a general TODO: look at `MutableArrayRef` to the code

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 12 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 11 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 13 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 15 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: The sloped cut constraint actually is contained within the left polytope. Note that due to them overlapping, I coudn't fully draw the left polytope. I think this is a limitation of the kind of representation and I will try to come up with a better example that is not limited like that.

Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I changed the example, hopefully, it is clearer now. I tried to do the same example as the integer set coalescing paper uses for the cut case.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 17 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: done

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 18 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: How do you mean this? I tried substituting this fucntion with llvm::to_vector, but this seems to take a size parameter and not an `ArrayRef.`

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 19 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: This cannot happen within my code, but as I define this for the whole mlir namespace, I decided to cover this case.

I will change this to an assertion, as someone who is actually comparing constraints with different sizes might be doing something else wrong.

Note: Depending on our representation of existential dimensions and int divisions, this case might happen actually AFAIU.

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 20 above)
Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: agreed and done

File: mlir/include/mlir/Analysis/Presburger/Simplex.h (hunk 8 above)
Comment (Dinistro) [reply mlir/include/mlir/Analysis/Presburger/Simplex.h]: I think the idea was to change ADJ_INEQ to AdjIneq

Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Simplex.h]: Ah okay, I thought generally enumerations are done in UPPERCASE, which doesn't seem to be the case as Direction is UpperCameCase now too. I will change ADJINEQ -> AdjIneq and ADJEQ -> AdjEq.

Comment (webmiche) [reply mlir/include/mlir/Analysis/Presburger/Simplex.h]: I just realiesed that I will have to do the same for the other Types two '^^

File: mlir/lib/Analysis/Presburger/Coalesce.cpp (hunk 18 above)
Comment (Dinistro) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: `to_vector` takes a `Range`. Note that this isn't per se a specific class as they extract begin and end with `std::begin` and `std::end`. 

https://llvm.org/doxygen/SmallVector_8h_source.html line 964

It seems like it also requires a `Size` template parameter that in you case would be 8.

Comment (webmiche) [reply mlir/lib/Analysis/Presburger/Coalesce.cpp]: I understood this now and use `llvm::to_vector`for this.

**Pull request closed without merging**